packages/xo-web/src/xo-app/vm/tab-network.js

```diff
@@ -304,6 +311,244 @@ class VifStatus extends BaseComponent {
+                </SingleLineRow>
+              )}
+              <SingleLineRow>
+                <Col size={6}>{_('aclRuleDirection')}</Col>
+                <Col size={6}>
+                  <Select
+                    name='direction'
+                    onChange={this.linkState('direction')}
+                    options={[
+                      { label: 'from', value: 'from' },
+                      { label: 'to', value: 'to' },
+                      { label: 'from/to', value: 'from/to' },
+                    ]}
+                    required
+                    value={direction}
+                  />
+                </Col>
+              </SingleLineRow>
+            </Container>
+          </div>
+        </fieldset>
+      </form>
+    )
+  }
+}
+
+@addSubscriptions({
+  plugins: subscribePlugins,
+})
+class AclRuleRow extends Component {
+  render() {
+    const { rule, vif, plugins } = this.props
+    const ruleObj = JSON.parse(rule)
+    const sdnControllerLoaded = plugins.some(
+      plugin => plugin.name === 'sdn-controller' && plugin.loaded
+    )
+
+    return (
+      <tr>
+        <td>{ruleObj.allow ? 'Enable' : 'Disable'}</td>
+        <td>{ruleObj.protocol}</td>
+        <td>{ruleObj.port}</td>
+        <td>{ruleObj.ipRange}</td>
+        <td>{ruleObj.direction}</td>
+        <td className='text-xs-right'>
+          <ActionRowButton
+            disabled={!sdnControllerLoaded}
+            handler={deleteAclRule}
+            handlerParam={{ ...ruleObj, vif }}
+            icon='delete'
+            level='danger'
+            tooltip={
+              sdnControllerLoaded
+                ? _('deleteRule')
+                : _('sdnControllerNotLoaded')
+            }
+          />
+        </td>
+      </tr>
+    )
+  }
+}
+
+@addSubscriptions({
+  plugins: subscribePlugins,
+})
+class AclRulesRows extends BaseComponent {
+  _newAclRule(vif) {
+    return confirm({
+      icon: 'add',
+      title: _('addRule'),
+      body: <NewAclRuleForm />,
+    }).then(({ allow, protocol, port, ipRange, direction }) => {
+      const hasProtocol = protocol != null
+      const usePort =
+        hasProtocol &&
+        USABLE_PORT_PROTOCOL.includes(protocol.value) &&
+        port !== undefined
+
+      return addAclRule({
+        allow,
+        protocol: hasProtocol ? protocol.value : undefined,
+        port: usePort ? +port : undefined,
+        ipRange: hasProtocol ? ipRange : '',
+        direction: direction.value,
+        vif,
+      })
+    })
+  }
+
+  _getRules = createSelector(
+    () => this.props.vif.other_config['xo:sdn-controller:of-rules'],
+    json => (json !== undefined ? JSON.parse(json) : undefined)
+  )
+
+  render() {
+    const { vif, plugins } = this.props
+    const { showRules } = this.state
+    const sdnControllerLoaded = plugins.some(
+      plugin => plugin.name === 'sdn-controller' && plugin.loaded
+    )
+    const rules = this._getRules()
+    const rulesToSee = rules !== undefined
```

Comment:
Should we show an empty table instead? Are there any logical differences between `xo:sdn-controller:of-rules` being `[]` or being `undefined`? If there aren't, the UI behaviour should the same.

Reply:
It can't be `[]`: either the `other-config[xo:sdn-controller:of-rules]` exists and it's a non-empty array, either its `undefined`.